ci/docker/linux-apt-python-313-freethreading.dockerfile

```diff
@@ -0,0 +1,51 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+ARG base
+FROM ${base}
+
+RUN apt install -y -q --no-install-recommends software-properties-common gpg-agent
+RUN add-apt-repository -y ppa:deadsnakes/ppa && apt update
+RUN apt install -y -q --no-install-recommends python3.13-dev python3.13-nogil python3.13-venv
+
+COPY python/requirements-build.txt \
+     python/requirements-test.txt \
+     /arrow/python/
+
+ENV ARROW_PYTHON_VENV /arrow-dev
+RUN python3.13t -m venv ${ARROW_PYTHON_VENV}
+RUN ${ARROW_PYTHON_VENV}/bin/python -m pip install -U pip setuptools wheel
+RUN ${ARROW_PYTHON_VENV}/bin/python -m pip install \
+      --pre \
+      --prefer-binary \
+      --extra-index-url "https://pypi.anaconda.org/scientific-python-nightly-wheels/simple" \
+      -r arrow/python/requirements-build.txt \
+      -r arrow/python/requirements-test.txt
+
+ENV ARROW_ACERO=ON \
+    ARROW_BUILD_STATIC=OFF \
+    ARROW_BUILD_TESTS=OFF \
+    ARROW_BUILD_UTILITIES=OFF \
+    ARROW_COMPUTE=ON \
+    ARROW_CSV=ON \
+    ARROW_DATASET=ON \
+    ARROW_FILESYSTEM=ON \
+    ARROW_GDB=ON \
+    ARROW_HDFS=ON \
+    ARROW_JSON=ON \
+    ARROW_USE_GLOG=OFF \
+    PYTHON_GIL=0
```

Comment:
Is the `PYTHON_GIL` environment variable used somewhere?

Reply:
It's used so that the test suite runs with the GIL disabled.

Reply:
But where is this env var taken into account?

Reply:
Ping @lysnikolaou

Reply:
When running the test suite, `cffi` is being imported. Because `cffi` has not declared support for free-threading yet, the GIL will be enabled, resulting in the tests running with the GIL enabled. If `PYTHON_GIL` has been set to `0`, the enabling of the GIL is prevented.

Reply:
So `PYTHON_GIL` is a cffi-specific environment variable?

Reply:
Also, this doesn't seem to have any effect:
```console
# export PYTHON_GIL=
# python3
Python 3.13.0rc2 experimental free-threading build (main, Sep  8 2024, 23:21:28) [GCC 11.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys
>>> sys._is_gil_enabled()
False
>>> import cffi
>>> sys._is_gil_enabled()
False
```

Reply:
Sorry for not making myself clear. No, `PYTHON_GIL` is a Python-related environment variable. Maybe [this excerpt from the Python docs](https://docs.python.org/3.13/whatsnew/3.13.html#free-threaded-cpython) would be better than me in trying to explain it?

> C-API extension modules need to be built specifically for the free-threaded build. Extensions that support running with the [GIL](https://docs.python.org/3.13/glossary.html#term-GIL) disabled should use the [Py_mod_gil](https://docs.python.org/3.13/c-api/module.html#c.Py_mod_gil) slot. Extensions using single-phase init should use [PyUnstable_Module_SetGIL()](https://docs.python.org/3.13/c-api/module.html#c.PyUnstable_Module_SetGIL) to indicate whether they support running with the GIL disabled. Importing C extensions that don’t use these mechanisms will cause the GIL to be enabled, unless the GIL was explicitly disabled with the [PYTHON_GIL](https://docs.python.org/3.13/using/cmdline.html#envvar-PYTHON_GIL) environment variable or the [-X gil=0](https://docs.python.org/3.13/using/cmdline.html#cmdoption-X) option.

Reply:
Ah, I see it's documented here: https://docs.python.org/3.14/using/cmdline.html#envvar-PYTHON_GIL

Google isn't very good at finding it...